app/router.go

```diff
@@ -102,63 +99,27 @@ func RegisterTopologyRoutes(router *mux.Router, r Reporter) {
 		gzipHandler(requestContextDecorator(makeProbeHandler(r))))
 }
 
-type byteCounter struct {
-	next  io.ReadCloser
-	count *uint64
-}
-
-func (c byteCounter) Read(p []byte) (n int, err error) {
-	n, err = c.next.Read(p)
-	*c.count += uint64(n)
-	return n, err
-}
-
-func (c byteCounter) Close() error {
-	return c.next.Close()
-}
-
 // RegisterReportPostHandler registers the handler for report submission
 func RegisterReportPostHandler(a Adder, router *mux.Router) {
 	post := router.Methods("POST").Subrouter()
 	post.HandleFunc("/api/report", requestContextDecorator(func(ctx context.Context, w http.ResponseWriter, r *http.Request) {
 		var (
-			rpt                              report.Report
-			reader                           = r.Body
-			err                              error
-			compressedSize, uncompressedSize uint64
+			rpt    report.Report
+			reader = r.Body
 		)
 
-		if log.GetLevel() == log.DebugLevel {
-			reader = byteCounter{next: reader, count: &compressedSize}
-		}
-		if strings.Contains(r.Header.Get("Content-Encoding"), "gzip") {
-			reader, err = gzip.NewReader(reader)
-			if err != nil {
-				respondWith(w, http.StatusBadRequest, err)
-				return
-			}
-		}
-
-		if log.GetLevel() == log.DebugLevel {
-			reader = byteCounter{next: reader, count: &uncompressedSize}
-		}
-		decoder := gob.NewDecoder(reader).Decode
+		gzipped := strings.Contains(r.Header.Get("Content-Encoding"), "gzip")
+		var handle codec.Handle
 		if strings.HasPrefix(r.Header.Get("Content-Type"), "application/json") {
-			decoder = codec.NewDecoder(reader, &codec.JsonHandle{}).Decode
+			handle = &codec.JsonHandle{}
 		} else if strings.HasPrefix(r.Header.Get("Content-Type"), "application/msgpack") {
-			decoder = codec.NewDecoder(reader, &codec.MsgpackHandle{}).Decode
+			handle = &codec.MsgpackHandle{}
 		}
 
-		if err := decoder(&rpt); err != nil {
+		if err := rpt.ReadBinary(reader, gzipped, handle); err != nil {
```

Comment:
`handle` must be non-null here, right? Otherwise things explode, probably with a rather obscure error. Perhaps we should enforce the non-nullness here by adding another branch to the preceding if/else. Or, even better, turn that into a `switch` with a default case that complains suitably.